src/Nethermind/Nethermind.Peering.Mothra/MothraLibp2p.cs

```diff
@@ -0,0 +1,240 @@
+                args.Add("--listen-address");
+                args.Add(settings.ListenAddress);
+            }
+
+            if (settings.MaximumPeers != null)
+            {
+                args.Add("--maxpeers");
+                args.Add(settings.MaximumPeers.ToString());
+            }
+
+            if (settings.BootNodes.Count > 0)
+            {
+                args.Add("--boot-nodes");
+                string value = string.Join(',', settings.BootNodes);
+                args.Add(value);
+            }
+
+            if (settings.Port != null)
+            {
+                args.Add("--port");
+                args.Add(settings.Port.ToString());
+            }
+
+            if (settings.DiscoveryPort != null)
+            {
+                args.Add("--discovery-port");
+                args.Add(settings.DiscoveryPort.ToString());
+            }
+
+            if (settings.DiscoveryAddress != null)
+            {
+                args.Add("--discovery-address");
+                args.Add(settings.DiscoveryAddress);
+            }
+
+            if (settings.Topics.Count > 0)
+            {
+                args.Add("--topics");
+                string value = string.Join(',', settings.Topics);
+                args.Add(value);
+            }
+
+            if (settings.PeerMultiAddresses.Count > 0)
+            {
+                args.Add("--libp2p-addresses");
+                string value = string.Join(',', settings.PeerMultiAddresses);
+                args.Add(value);
+            }
+
+            if (settings.DebugLevel != null)
+            {
+                if (_debugLevels.TryGetValue(settings.DebugLevel.Value, out string value))
+                {
+                    args.Add("--debug-level");
+                    args.Add(value);
+                }
+            }
+
+            if (settings.VerbosityLevel != null)
+            {
+                args.Add("--verbosity");
+                args.Add(settings.VerbosityLevel.ToString());
+            }
+
+            return args.ToArray();
+        }
+
+        private unsafe void DiscoveredPeerHandler(byte* peerUtf8Ptr, int peerLength)
+        {
+            string peer = new String((sbyte*) peerUtf8Ptr, 0, peerLength, Encoding.UTF8);
+            OnPeerDiscovered(new PeerDiscoveredEventArgs(peer));
+        }
+
+        private void OnGossipReceived(GossipReceivedEventArgs e)
+        {
+            GossipReceived?.Invoke(this, e);
+        }
+
+        private void OnPeerDiscovered(PeerDiscoveredEventArgs e)
+        {
+            PeerDiscovered?.Invoke(this, e);
+        }
+
+        private void OnRpcReceived(RpcReceivedEventArgs e)
+        {
+            RpcReceived?.Invoke(this, e);
+        }
+
+        private unsafe void ReceiveGossipHandler(byte* topicUtf8Ptr, int topicLength, byte* dataPtr, int dataLength)
+        {
+            string topic = new String((sbyte*) topicUtf8Ptr, 0, topicLength, Encoding.UTF8);
+            byte[] data = new byte[dataLength];
```

Comment:
can probably do it in a non-allocating way by Unsafe.Add to topic ptr and comparing with cached topics.